File: JS/coin-change.js
function coinChange(coins, amount) {
  const dp = new Array(amount + 1).fill(Infinity);
  dp[0] = 0;

  for (let i = 1; i <= amount; i++) {
    for (const coin of coins) {
      if (coin <= i && dp[i - coin] + 1 < dp[i]) {
        dp[i] = dp[i - coin] + 1;
      }
    }
  }

  return dp[amount] === Infinity ? -1 : dp[amount];
}

const lines = require('fs').readFileSync('/dev/stdin', 'utf8').split('\n');
const numCoins = parseInt(lines[0]);
const coins = lines[1].trim().split(/\s+/).map(Number).slice(0, numCoins);
const amount = parseInt(lines[2]);
console.log(coinChange(coins, amount));

File: C/coin-change.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int main() {
    int numCoins;
    scanf("%d", &numCoins);
    int *coins = malloc(sizeof(int) * numCoins);
    for (int i = 0; i < numCoins; i++) scanf("%d", &coins[i]);
    int amount;
    scanf("%d", &amount);

    int *dp = malloc(sizeof(int) * (amount + 1));
    dp[0] = 0;
    for (int i = 1; i <= amount; i++) dp[i] = INT_MAX;

    for (int i = 1; i <= amount; i++) {
        for (int j = 0; j < numCoins; j++) {
            int coin = coins[j];
            if (coin <= i && dp[i - coin] != INT_MAX) {
                if (dp[i - coin] + 1 < dp[i]) dp[i] = dp[i - coin] + 1;
            }
        }
    }

    printf("%d\n", dp[amount] == INT_MAX ? -1 : dp[amount]);
    free(coins);
    free(dp);
    return 0;
}